documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1182 @@
+
+Aside from the api below, the following constraints should also be followed:
+- Three Dimensional Vectors should implement a CrossProduct function which takes another vector and returns the cross product with our original vector. A static implementation of this function should be available as well.
+- Every Dimension should have the relevant number of properties (X and Y for Vector2, and X,Y,Z,W,V for Vector5) and relevant unit vectors
+- Constructors for higher dimensions should include lower dimension variants that use the lower dimensions for their specific components (vector2 -> X,Y)
+
+```cs
+public readonly record struct Vector2I<TScalar> : IEquatable<Vector2I<TScalar>>, 
+                           IEquatable<Vector2>,
+                           IBinaryIntegerVector<Vector2I<TScalar>, TScalar>,
+                           IVector2<Vector2I<TScalar>, TScalar>
+                           where TScalar : IBinaryInteger<TScalar>
+{
+    ///<summary>
+    ///The X component of the Vector
+    ///</summary>
+    [DataMember]
+    public TScalar X { get; }
+
+    ///<summary>
+    ///The Y component of the Vector
+    ///</summary>
+    [DataMember]
+    public TScalar Y { get; }
+
+    public static Vector2I<TScalar> UnitX { get; }
+    public static Vector2I<TScalar> UnitY { get; }
+
+    public static Vector2I<TScalar> Zero { get; }
+    public static Vector2I<TScalar> One { get; }
+
+    public TScalar this[int index] { get; }
+
+    public Vector2I(TScalar x, TScalar y);
+
+    public static Vector2I<TScalar> operator +(Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> operator -(Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> operator *(Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> operator /(Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> operator %(Vector2I<TScalar> left, Vector2I<TScalar> right);
+
+    public static Vector2I<TScalar> operator *(Vector2I<TScalar> left, TScalar right);
+    public static Vector2I<TScalar> operator /(Vector2I<TScalar> left, TScalar right);
+    public static Vector2I<TScalar> operator %(Vector2I<TScalar> left, TScalar right);
+
+    public static Vector2I<TScalar> operator *(TScalar left, Vector2I<TScalar> right);
+
+    public static Vector2I<TScalar> operator +(Vector2I<TScalar> value);
+    public static Vector2I<TScalar> operator -(Vector2I<TScalar> value);
+
+    public static Vector2I<TScalar> operator &(Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> operator |(Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> operator ^(Vector2I<TScalar> left, Vector2I<TScalar> right);
+
+    public static Vector2I<TScalar> operator ~(Vector2I<TScalar> value);
+
+    public TScalar LengthSquared { get; }
```

Comment:
This may be a method or property, we haven't decided yet.

Reply:
I feel like Resharper assuming length refers to the number of elements is its problem, context matters. Similarly if we wrote a LineSegment class we would probably have a length property, and in context it makes sense to call it that. The length of a vector is generally considered the length mathematically. Length matches the 2.X and System.Numerics naming.

If we feel we need to rename Unreal uses Size for this property, while unity uses Magnitude, either would be good alternatives.

Reply:
An automated tool only makes this assumption because a human thought that it was a safe assumption to make. The same could be true for any of our users. Properties don’t elicit the idea of performing a calculation, rather accessing something that already exists about the value.

I’m fine with length if it’s a method per this argument. However I’d also be open to calling it something like magnitude (size would be confusing as well) if we do want to keep it as a property.

Reply:
I think assuming a user with context would make the same assumptions as a tool which has none is a bit of a stretch. For example, a length property on a Rectangle would not be assumed to relate to number of elements. Similarly, I don't think my first assumption with a Vector's Length is that it relates to number of elements as vectors since Mathematics gives meaning to the length of a vector.

Properties also do calculations all the time. For example, the IsIdentity property in matricies requires several comparisons. Properties are by definition obfuscation of accessor and mutator methods which may require any number of calculations or checks.

Also does a Length method not get picked up by Resharper?

I do agree that Size is a bad name.

Reply:
Also the Length property should have a summary which tells a user that it refers to the Euclidean Length of the Vector.

Reply:
Its going to be a property in the generic ones for System.Numerics.

Length is the industry standard term (more so than magnitude) and it is an incredibly cheap computation